# What's in an Avocado Toast: A Supply Chain Analysis

![](avocado_wallpaper.jpeg)

You find yourself in London, crafting a delectable avocado toast, a dish that has risen dramatically in popularity on breakfast menus since the 2010s. This straightforward recipe requires just five ingredients: a ripe avocado, half a lemon, a generous pinch of salt flakes, two slices of sourdough bread, and a good drizzle of extra virgin olive oil. Most of these ingredients are now staples in grocery stores, and as you will find with this project, that is no small feat!

In this project, you'll conduct a supply chain analysis of three ingredients used in avocado toast using the Open Food Facts database. This database contains extensive, openly-sourced information on various foods, including their origins. Through this analysis, you will gain an in-depth understanding of the complex supply chain involved in producing a single dish.

Three pairs of files are provided in the data folder:
- A CSV file for each ingredient, such as `avocado.csv`, with data about each food item and countries of origin.
- A TXT file for each ingredient, such as `relevant_avocado_categories`, containing only the category tags of interest for that food.

Here are some other key points about these files:
- Some of the rows of data in each of the three CSV files do not contain relevant data for your investigation. In each dataset, you will need to filter out rows with irrelevant data, based on values in the `categories_tags` column. Examples of categories are fruits, vegetables, and fruit-based oils. Filter the DataFrame to include only rows where `categories_tags` contains one of the tags in the relevant categories for that ingredient.
- Each row of data usually has multiple category tags in the `categories_tags` column. There is a column in each CSV file called `origins_tags`, which contains strings for the country of origin of each item.

After completing this project, you'll be armed with a list of ingredients and their countries of origin and be well-positioned to launch into other analyses that explore how long, on average, these ingredients spend at sea.

[Open Food Facts database](https://world.openfoodfacts.org/)

In [1]:
#import packages
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [2]:
#load in the data

#CSV's
avocado_orig = pd.read_csv("avocado.csv", sep="\t")
lemon_orig = pd.read_csv("lemon.csv", delimiter="\t")
olive_oil_orig = pd.read_csv("olive_oil.csv", delimiter="\t")
salt_flakes_orig = pd.read_csv("salt_flakes.csv", delimiter="\t")
sourdough_orig = pd.read_csv("sourdough.csv", delimiter="\t")

#TXT's
avocado_cats = pd.read_csv("relevant_avocado_categories.txt", header=None).rename(columns={0:"category"})
olive_oil_cats = pd.read_csv("relevant_olive_oil_categories.txt", header=None).rename(columns={0:"category"})
sourdough_cats = pd.read_csv("relevant_sourdough_categories.txt", header=None).rename(columns={0:"category"})

display(avocado_orig, avocado_cats)


C:\Users\elija\AppData\Local\Temp\ipykernel_32208\3018457581.py:5: DtypeWarning: Columns (1,2,4,5,7,8,9,10,12,15,17,18,19,21,22,23,24,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,43,44,46,47,48,49,50,51,52,53,54,55,75,76,77,80,81,82,83,84,85,86,88,90,91,93,94,95,97,98,99,100,102,103,104,105,106,107,108,109,120,124,128,132,136,140,144,148,152,156,162,164,166,168,170,174,178,180,182,184,186,188,190,200,202,204,206,208,210,214,216,220,222,224,242,256,258,260,262,264,266,270,272,274,278,280,281,282,283,284,285,286,287,288,289,290,292,293,294,295,296,307,321,322,323,324,325,326) have mixed types. Specify dtype option on import or set low_memory=False.
  lemon_orig = pd.read_csv("lemon.csv", delimiter="\t")
C:\Users\elija\AppData\Local\Temp\ipykernel_32208\3018457581.py:6: DtypeWarning: Columns (0,1,2,4,5,6,7,8,10,13,15,16,17,18,20,21,22,23,24,25,26,27,29,30,31,32,33,34,35,36,37,38,39,42,43,44,45,46,48,49,50,57,67,70,71,72,73,74,75,76,79,80,81,82,83,86,87,88,89,90,91,92,95,96,97,98,99,100

,code,lc,product_name_de,product_name_el,product_name_en,product_name_es,product_name_fi,product_name_fr,product_name_id,product_name_it,...,off:ecoscore_data.adjustments.packaging.non_recyclable_and_non_biodegradable_materials,off:ecoscore_data.adjustments.production_system.value,off:ecoscore_data.adjustments.threatened_species.value,sources_fields:org-database-usda:available_date,sources_fields:org-database-usda:fdc_category,sources_fields:org-database-usda:fdc_data_source,sources_fields:org-database-usda:fdc_id,sources_fields:org-database-usda:modified_date,sources_fields:org-database-usda:publication_date,data_sources
0,0059749979702,fr,NaN,NaN,NaN,NaN,NaN,Naturalia Avocado Oil,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"App - yuka, Apps"
1,7610095131409,en,NaN,NaN,NaN,NaN,NaN,Avocado Bowl chips,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"App - Yuka, Apps, Producers, Producer - zweifel"
2,4005514005578,en,NaN,NaN,Gelbe Linse Avocado Brotaufstrich,NaN,NaN,NaN,NaN,NaN,...,1.0,15.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"App - yuka, Apps, App - smoothie-openfoodfacts"
3,0879890002513,en,NaN,NaN,Avocado toast chili lime,NaN,NaN,NaN,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"App - Yuka, Apps, App - InFood"
4,0223086613685,en,NaN,NaN,Avocado,NaN,NaN,NaN,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"App - Yuka, Apps"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1780,0819573012712,en,NaN,NaN,"Organic Baby Food, Apples, Kale & Avocados",NaN,NaN,NaN,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Database - USDA NDB, Databases"
1781,0052200072097,en,NaN,NaN,"Just Pineapple, Pear & Avocado",NaN,NaN,NaN,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Database - USDA NDB, Databases"
1782,0793613300000,en,NaN,NaN,Spinach Avocado Dip,NaN,NaN,NaN,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Database - USDA NDB, Databases"
1783,05252428,en,NaN,NaN,"Organic Just Apple, Raspberry & Avocado, Apple...",NaN,NaN,NaN,NaN,NaN,...,1.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Database - USDA NDB, Databases"


,category
0,en:avocadoes
1,en:avocados
2,en:fresh-foods
3,en:fresh-vegetables
4,en:fruchte
5,en:fruits
6,en:raw-green-avocados
7,en:tropical-fruits
8,en:tropische-fruchte
9,en:vegetables-based-foods


Shapes of the imported datasets:
- `avocado_orig`: 1785 (rows) x 184 (columns)
- `lemon_orig`: 8107 x 328
- `olive_oil_orig`: 8298 x 326
- `salt_flakes_orig`: 54 x 104
- `sourdough_orig`: 1422 x 189
- `avocado_cats`: 11 x 1
- `olive_oil_cats`: 108 x 1
- `sourdough_cats`: 33 x 1

# Objectives
Firstly, apply information given earlier:
- Some of the rows of data in each of the three CSV files do not contain relevant data for your investigation. In each dataset, you will need to filter out rows with irrelevant data, based on values in the `categories_tags` column. Examples of categories are fruits, vegetables, and fruit-based oils. Filter the DataFrame to include only rows where `categories_tags` contains one of the tags in the relevant categories for that ingredient.
- Each row of data usually has multiple category tags in the `categories_tags` column.



Apply your data manipulation and analysis skills to investigate the supply chain of ingredients for making avocado toast in the U.K. Your task is to determine the following information:
- The name of the most common country of origin for each of the three key ingredients: avocados, olive oil, and sourdough.

Store the most common country of origin for each ingredient in the following variables: `top_avocado_origin`, `top_olive_oil_origin`, and `top_sourdough_origin`. Ensure that the country names contain only letters (A-Z) and spaces, with no hyphens or other characters.

To focus your analysis, subset each of the DataFrames to include only these relevant columns: 'code', 'lc', 'product_name_en', 'quantity', 'serving_size', 'packaging_tags', 'brands', 'brands_tags', 'categories_tags', 'labels_tags', 'countries', 'countries_tags', 'origins', 'origins_tags'.

**Notes:**
- Want to filter the 'categories_tags' variable in each respective dataset (.csv) file for observations that contain at least 1 of the relevant categories as listed in the corresponding .txt file. However, since many observations have multiple categories in the 'categories_tags' variable, which are separated by commas, need to be able to search through each observation. Can do this by putting each 'categories_tags' observation into a list.
    - Need to do this process for the avocado, olive-oil, & sourdough datasets.

## Cleaning & Filtering
### Avocado
Recall that the original shape is 1785 rows x 184 columns.

In [3]:
#drop null 'categories_tags' observations from avocado data
avocado_I = avocado_orig[avocado_orig.categories_tags.notna()].reset_index(drop=True)
    #display(len(avocado_orig), len(avocado))

#make a new column for 'categories_tags' that is a list of each observation
avocado_I['categories_list'] = avocado_I.categories_tags.apply(lambda x: x.split(","))
#display(avocado_I[['code','categories_tags','categories_list']])

#filter for observations which have 'categories_tags' in the .txt categories
    #for each row, have to search through each element in 'category_list' & see if they match any element in corresponding .txt
avocado_I['in_cats?'] = avocado_I.categories_list.apply(lambda x: any(i in [*avocado_cats.category] for i in x))
#display(avocado_I[['code','categories_list','in_cats?']])

#using this new column, 'in_cats?', filter for rows that have a category tag ('categories_tags'/'categories_list') in 
    #the relevant category tags specified in the corresponding .txt file
avocado = avocado_I[avocado_I['in_cats?'] == True].reset_index(drop=True)
    #are 182 such rows

#subset for main relevant columns, as specified in Objectives
avocado = avocado[['code', 'lc', 'product_name_en', 'quantity', 'serving_size', 'packaging_tags', 'brands', 'brands_tags', 
                   'categories_tags', 'labels_tags', 'countries', 'countries_tags', 'origins', 'origins_tags']]
display(avocado)


,code,lc,product_name_en,quantity,serving_size,packaging_tags,brands,brands_tags,categories_tags,labels_tags,countries,countries_tags,origins,origins_tags
0,3662994002063,fr,NaN,3 fruits,NaN,NaN,la compagnie des fruits mûrs,la-compagnie-des-fruits-murs,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,France,en:france,NaN,NaN
1,8437013031011,fr,NaN,1 kg,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,France,en:france,NaN,NaN
2,4016249238155,de,NaN,135g,100g,de:gläschen,Allos,allos,"en:plant-based-foods-and-beverages,en:plant-ba...","en:organic,en:vegetarian,en:eu-organic,en:no-g...",Deutschland,en:germany,Europäische Union,en:european-union
3,8718963381532,de,NaN,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,Deutschland,en:germany,NaN,NaN
4,8436002746707,es,NaN,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,España,en:spain,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
177,3700353611218,fr,NaN,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,France,en:france,NaN,NaN
178,4311527575718,de,NaN,NaN,NaN,de:aufkleber,Edeka,edeka,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,Deutschland,en:germany,Peru,en:peru
179,4311527571499,en,NaN,NaN,NaN,en:aufkleber,Edeka,edeka,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,Germany,en:germany,NaN,NaN
180,3439496511399,fr,NaN,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,France,en:france,NaN,NaN


### Olive Oil
Recall that the original shape is 8,298 rows x 326 columns.

In [4]:
#drop null 'categories_tags' observations from olive-oil data
olive_oil_I = olive_oil_orig[olive_oil_orig.categories_tags.notna()].reset_index(drop=True)
    #display(len(olive_oil_orig), len(olive_oil))

#make a new column for 'categories_tags' that is a list of each observation
olive_oil_I['categories_list'] = olive_oil_I.categories_tags.apply(lambda x: x.split(","))
#display(olive_oil_I[['code','categories_tags','categories_list']])

#filter for observations which have 'categories_tags' in the .txt categories
    #for each row, have to search through each element in 'category_list' & see if they match any element in corresponding .txt
olive_oil_I['in_cats?'] = olive_oil_I.categories_list.apply(lambda x: any(i in [*olive_oil_cats.category] for i in x))
#display(olive_oil_I[['code','categories_list','in_cats?']])

#using this new column, 'in_cats?', filter for rows that have a category tag ('categories_tags'/'categories_list') in 
    #the relevant category tags specified in the corresponding .txt file
olive_oil = olive_oil_I[olive_oil_I['in_cats?'] == True].reset_index(drop=True)
    #are 5,290 such rows

#subset for main relevant columns, as specified in Objectives
olive_oil = olive_oil[['code', 'lc', 'product_name_en', 'quantity', 'serving_size', 'packaging_tags', 'brands', 'brands_tags', 
                   'categories_tags', 'labels_tags', 'countries', 'countries_tags', 'origins', 'origins_tags']]
display(olive_oil)


,code,lc,product_name_en,quantity,serving_size,packaging_tags,brands,brands_tags,categories_tags,labels_tags,countries,countries_tags,origins,origins_tags
0,3560070910366,fr,NaN,1 l,NaN,"en:bottle,fr:bouteille-verre-de-6g,fr:bouteill...","Carrefour Bio,Carrefour","carrefour-bio,carrefour","en:plant-based-foods-and-beverages,en:plant-ba...","en:organic,en:eu-organic,en:non-eu-agriculture...","France, Pologne, Roumanie","en:france,en:poland,en:romania","Espagne, Tunisie","en:spain,en:tunisia"
1,3270190008262,fr,NaN,1 l,NaN,"en:glass,en:bottle,fr:bouchon-metallique,fr:bo...",Carrefour,carrefour,"en:plant-based-foods-and-beverages,en:plant-ba...",en:green-dot,"France, Guadeloupe, Roumanie","en:france,en:guadeloupe,en:romania",Union européenne,en:european-union
2,6191509903023,en,Huile d'olive vierge extra,1 l,NaN,"en:glass,en:bottle,fr:bouchon-metal",Terra Delyssa,terra-delyssa,"en:plant-based-foods-and-beverages,en:plant-ba...","en:organic,en:certified-by-ecocert,en:eu-organ...","France, United States","en:france,en:united-states",Tunisia,en:tunisia
3,6191509900855,fr,NaN,750 ml,NaN,"en:glass,en:bottle",Terra Delyssa,terra-delyssa,"en:plant-based-foods-and-beverages,en:plant-ba...","en:organic,en:certified-by-ecocert,en:eu-organ...","France, Suisse","en:france,en:switzerland",Tunisie,en:tunisia
4,8002470023154,fr,NaN,"0,75 cl",NaN,"fr:bouchon-en-plastique-et-caoutchouc,fr:boute...",Carapelli,carapelli,"en:plant-based-foods-and-beverages,en:plant-ba...","en:organic,en:eu-organic,en:non-eu-agriculture...",France,en:france,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5285,695305,en,Olive oil with white truffle,NaN,NaN,NaN,Trader Joe's,trader-joe-s,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,United States,en:united-states,NaN,NaN
5286,6009681271000,en,Extra Virgin olive oil,NaN,5.0g,NaN,Willow creek,willow-creek,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,South Africa,en:south-africa,NaN,NaN
5287,4012625536136,fr,NaN,NaN,NaN,NaN,Marmara,marmara,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,France,en:france,NaN,NaN
5288,8426622284000,en,Olive oil,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,United Kingdom,en:united-kingdom,NaN,NaN


### Sourdough
Recall that the original shape is 1,422 rows x 189 columns.

In [5]:
#drop null 'categories_tags' observations from sourdough data
sourdough_I = sourdough_orig[sourdough_orig.categories_tags.notna()].reset_index(drop=True)
    #display(len(sourdough_orig), len(sourdough))

#make a new column for 'categories_tags' that is a list of each observation
sourdough_I['categories_list'] = sourdough_I.categories_tags.apply(lambda x: x.split(","))
#display(sourdough_I[['code','categories_tags','categories_list']])

#filter for observations which have 'categories_tags' in the .txt categories
    #for each row, have to search through each element in 'category_list' & see if they match any element in corresponding .txt
sourdough_I['in_cats?'] = sourdough_I.categories_list.apply(lambda x: any(i in [*sourdough_cats.category] for i in x))
#display(sourdough_I[['code','categories_list','in_cats?']])

#using this new column, 'in_cats?', filter for rows that have a category tag ('categories_tags'/'categories_list') in 
    #the relevant category tags specified in the corresponding .txt file
sourdough = sourdough_I[sourdough_I['in_cats?'] == True].reset_index(drop=True)
    #are 399 such rows

#subset for main relevant columns, as specified in Objectives
sourdough = sourdough[['code', 'lc', 'product_name_en', 'quantity', 'serving_size', 'packaging_tags', 'brands', 'brands_tags', 
                   'categories_tags', 'labels_tags', 'countries', 'countries_tags', 'origins', 'origins_tags']]
display(sourdough)


,code,lc,product_name_en,quantity,serving_size,packaging_tags,brands,brands_tags,categories_tags,labels_tags,countries,countries_tags,origins,origins_tags
0,10500016941075200179,fr,NaN,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,France,en:france,NaN,NaN
1,10500016930534800159,fr,NaN,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,Royaume-Uni,en:united-kingdom,NaN,NaN
2,10086988,en,Brown Ciabatta Rolls with Buckwheat and Sourdough,200g,50g,en:plastic,Schär,schar,"en:plant-based-foods-and-beverages,en:plant-ba...","en:vegetarian,en:no-gluten,en:source-of-fibre,...","Belgium, France, Germany, Ireland, Italy, Spai...","en:belgium,en:france,en:germany,en:ireland,en:...",NaN,NaN
3,00252751,en,White sourdough,400g,100g,en:ldpe-bag,Sainsbury's,sainsbury-s,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,"France, United Kingdom","en:france,en:united-kingdom",NaN,NaN
4,5059697263682,en,Sundried Tomato & Oregano Bread with Sourdough,NaN,NaN,NaN,Tesco,tesco,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,United Kingdom,en:united-kingdom,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
394,0028833050306,en,Sprouted Sourdough Bread,NaN,1 SLICE (34 g),NaN,Alvarado St. Bakery,alvarado-st-bakery,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,United States,en:united-states,NaN,NaN
395,0076793300405,en,"The bread garden, northwest sourdough bread",NaN,1 SLICE (43 g),NaN,The Bread Garden,the-bread-garden,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,United States,en:united-states,NaN,NaN
396,26323466,en,Artisan Style Sourdough Multigrain,1 loaf,NaN,NaN,Bakers Life,bakers-life,"en:plant-based-foods-and-beverages,en:plant-ba...","en:australian-made,en:no-added-sugar",Australia,en:australia,NaN,NaN
397,4088600071985,en,Sourdough crumpets,NaN,NaN,NaN,NaN,NaN,"en:plant-based-foods-and-beverages,en:plant-ba...",NaN,United Kingdom,en:united-kingdom,NaN,NaN


## Analysis
*Recall:* Apply your data manipulation and analysis skills to investigate the supply chain of ingredients for making avocado toast in the U.K. Your task is to determine the following information:
- The name of the most common country of origin for each of the three key ingredients: avocados, olive oil, and sourdough.

Store the most common country of origin for each ingredient in the following variables: `top_avocado_origin`, `top_olive_oil_origin`, and `top_sourdough_origin`. Ensure that the country names contain only letters (A-Z) and spaces, with no hyphens or other characters.

**Notes:**
- Recall that the `origins_tags` variable refers to the country of origin for each item.
- Before counting, want to filter for observations from the UK. Can use the `countries` variable to do so.

In [23]:
## AVOCADO - after filtering & cleaning, has 182 rows
display("AVOCADO:")
display(avocado[avocado.countries == "United Kingdom"].origins_tags.value_counts(dropna=False))
display("--------------------------------------------------------------")

#OLIVE OIL - after filtering & cleaning, has 5,290 rows
display("OLIVE OIL:")
display(olive_oil[olive_oil.countries == "United Kingdom"].origins_tags.value_counts(dropna=False))
display("--------------------------------------------------------------")

#SOURDOUGH - after filtering & cleaning, has 399 rows
display("SOURDOUGH:")
display(sourdough[sourdough.countries == "United Kingdom"].origins_tags.value_counts(dropna=False))

'AVOCADO:'

origins_tags
NaN                 8
en:peru             2
en:spain,en:peru    1
en:chile,en:peru    1
en:israel           1
Name: count, dtype: int64

'--------------------------------------------------------------'

'OLIVE OIL:'

origins_tags
NaN                                                   127
en:greece                                               6
en:spain                                                4
en:italy                                                4
en:greece,en:italy,en:portugal,en:spain,en:tunisia      2
en:produce-of-italy                                     1
en:european-union-and-non-european-union                1
en:produced-in-italy                                    1
en:european-union                                       1
Name: count, dtype: int64

'--------------------------------------------------------------'

'SOURDOUGH:'

origins_tags
NaN                  37
en:united-kingdom     3
en:france             1
Name: count, dtype: int64

In [24]:
#assign the specified variables
top_avocado_origin = avocado[avocado.countries == "United Kingdom"].origins_tags.value_counts(
    dropna=True).index[0].split(":")[1].title()
top_olive_oil_origin = olive_oil[olive_oil.countries == "United Kingdom"].origins_tags.value_counts(
    dropna=True).index[0].split(":")[1].title()
top_sourdough_origin = sourdough[sourdough.countries == "United Kingdom"].origins_tags.value_counts(
    dropna=True).index[0].split(":")[1].replace("-", " ").title()

display("Most common country of origin for avocados: {}.".format(top_avocado_origin))
display("Most common country of origin for olive oil: {}.".format(top_olive_oil_origin))
display("Most common country of origin for sourdough: {}.".format(top_sourdough_origin))


'Most common country of origin for avocados: Peru.'

'Most common country of origin for olive oil: Greece.'

'Most common country of origin for sourdough: United Kingdom.'